cpp/src/barretenberg/proof_system/composer/permutation_helper.hpp

```diff
@@ -120,63 +127,95 @@ std::vector<CyclicPermutation> compute_wire_copy_cycles(const CircuitConstructor
+            if (generalized) {
+                mapping.ids[i].emplace_back(permutation_subgroup_element{
+                    .row_index = (uint32_t)j, .column_index = (uint8_t)i, .is_public_input = false, .is_tag = false });
+            }
         }
     }
 
+    // Represents the index of a variable in circuit_constructor.variables (needed only for generalized)
+    std::span<const uint32_t> real_variable_tags = circuit_constructor.real_variable_tags;
+
     // Go through each cycle
-    for (size_t i = 0; i < wire_copy_cycles.size(); ++i) {
-        for (size_t j = 0; j < wire_copy_cycles[i].size(); ++j) {
-            // Get the indices of the current node and next node int he cycle
-            cycle_node current_cycle_node = wire_copy_cycles[i][j];
+    size_t cycle_index = 0;
+    for (auto& copy_cycle : wire_copy_cycles) {
+        for (size_t node_idx = 0; node_idx < copy_cycle.size(); ++node_idx) {
+            // Get the indices of the current node and next node in the cycle
+            cycle_node current_cycle_node = copy_cycle[node_idx];
             // If current node is the last one in the cycle, then the next one is the first one
-            size_t next_cycle_node_index = j == wire_copy_cycles[i].size() - 1 ? 0 : j + 1;
-            cycle_node next_cycle_node = wire_copy_cycles[i][next_cycle_node_index];
+            size_t next_cycle_node_index = (node_idx == copy_cycle.size() - 1 ? 0 : node_idx + 1);
+            cycle_node next_cycle_node = copy_cycle[next_cycle_node_index];
             const auto current_row = current_cycle_node.gate_index;
             const auto next_row = next_cycle_node.gate_index;
 
-            const uint32_t current_column = current_cycle_node.wire_index;
-            const uint32_t next_column = next_cycle_node.wire_index;
+            const auto current_column = current_cycle_node.wire_index;
+            const auto next_column = static_cast<uint8_t>(next_cycle_node.wire_index);
             // Point current node to the next node
-            sigma_mappings[current_column][current_row] = {
-                .row_index = next_row, .column_index = (uint8_t)next_column, .is_public_input = false, .is_tag = false
+            mapping.sigmas[current_column][current_row] = {
+                .row_index = next_row, .column_index = next_column, .is_public_input = false, .is_tag = false
             };
+
+            if (generalized) {
+                bool first_node = (node_idx == 0);
+                bool last_node = (next_cycle_node_index == 0);
+
+                if (first_node) {
+                    mapping.ids[current_column][current_row].is_tag = true;
+                    mapping.ids[current_column][current_row].row_index = (real_variable_tags[cycle_index]);
+                }
+                if (last_node) {
+                    mapping.sigmas[current_column][current_row].is_tag = true;
+
+                    // TODO: yikes, std::maps are expensive. Can we find a way to get rid of this?
```

Comment:
Can you attribute this quote to Zac?

Reply:
I checked at some point that he wrote it.

Reply:
Thought about doing that. Will do